Use standard approach of displaying filters for list tables

merkushin opened the pull request:
In this PR I'm trying to display filters as it is done for Posts. It helps to avoid issues with positioning of elements. However, I found we have some special cases (the standard approach is to hide filters on mobile, but we have some required filters that we have to display always).

### Known Issues

* I noticed that filters started adding unnecessary params like `_wpnonce` and `_wp_http_referer` to the query. These params we add to the form, the same on the Posts page. But when I submit filters for Posts it doesn't add mentioned params to the query. Can't find how it is filtered there. Any ideas? 🙏 

### Changes proposed in this Pull Request

* Enclose the table, search box and filters into single form.

### Testing instructions

* Try to open Reports, Grading, Students pages.
* Compare them to `trunk`. There might be changes, but they should be "positive" (i.e. it should look like in Pages now instead of "custom" positioning).
* Check them with mobile dimensions.

### Screenshot / Video

*After*
<img width="1606" alt="CleanShot 2022-05-29 at 17 11 32@2x" src="https://user-images.githubusercontent.com/329356/170873535-f2eb91c3-dc1d-4667-8cc8-8ad30ae80b4c.png">
<img width="290" alt="CleanShot 2022-05-29 at 17 11 51@2x" src="https://user-images.githubusercontent.com/329356/170873546-cbc6fad4-28ba-4fa3-9242-83474e491eb0.png">
<img width="1280" alt="CleanShot 2022-05-24 at 10 52 01@2x" src="https://user-images.githubusercontent.com/329356/169979189-1546191e-61bc-4f8d-80e7-f0d32885a415.png">
<img width="319" alt="CleanShot 2022-05-24 at 10 52 19@2x" src="https://user-images.githubusercontent.com/329356/169979220-935d54a6-6201-4daa-862e-a8c3cff8c0ac.png">
<img width="1280" alt="CleanShot 2022-05-24 at 10 52 47@2x" src="https://user-images.githubusercontent.com/329356/169979246-6a1ca838-8d45-4013-85c1-601361dc8fc0.png">
<img width="322" alt="CleanShot 2022-05-24 at 10 53 02@2x" src="https://user-images.githubusercontent.com/329356/169979310-c80efde8-1eff-4030-8b1f-23cec7d3aa23.png">
<img width="806" alt="CleanShot 2022-05-29 at 17 29 55@2x" src="https://user-images.githubusercontent.com/329356/170874557-f319fce8-d3d6-44ac-bcc9-f19d423d05df.png">


*Before*
<img width="1279" alt="CleanShot 2022-05-24 at 10 56 34@2x" src="https://user-images.githubusercontent.com/329356/169980128-d7e56c08-488f-46b2-98f5-fe466245b3bc.png">
<img width="320" alt="CleanShot 2022-05-24 at 10 56 48@2x" src="https://user-images.githubusercontent.com/329356/169980206-4981cb6b-cd53-4012-9c1a-ff03692c39f5.png">
<img width="1280" alt="CleanShot 2022-05-24 at 10 57 11@2x" src="https://user-images.githubusercontent.com/329356/169980243-b0c00253-fa2f-48e1-a357-268ca4dcdffc.png">
<img width="321" alt="CleanShot 2022-05-24 at 10 57 26@2x" src="https://user-images.githubusercontent.com/329356/169980325-5e813586-3cde-4812-9147-ac9e2da0d41c.png">
<img width="1280" alt="CleanShot 2022-05-24 at 10 57 52@2x" src="https://user-images.githubusercontent.com/329356/169980372-95fb26e9-fb90-4e62-9f77-1698425e9e80.png">
<img width="321" alt="CleanShot 2022-05-24 at 10 58 08@2x" src="https://user-images.githubusercontent.com/329356/169980422-da52bd64-edad-4336-8b6d-30f934177e72.png">

merkushin committed:
Use standard approach of displaying filters for list tables

merkushin committed:
Revert commenting of position property

merkushin commented on includes/admin/views/html-admin-page-students-course.php:
```diff
@@ -20,16 +20,13 @@
+	<?php $sensei_list_table->table_search_form(); ?>
+	<?php $sensei_list_table->display(); ?>
+</form>
+<?php do_action( 'sensei_learners_extra' ); ?>
```
found a usage in Content Drip

merkushin commented on includes/admin/views/html-admin-page-students-main.php:
```diff
@@ -27,14 +27,13 @@
+				<?php $sensei_list_table->table_search_form(); ?>
+				<?php $sensei_list_table->display(); ?>
+			</form>
 			<?php do_action( 'sensei_learner_admin_extra' ); ?>
```
No usages found on GH

merkushin commented on includes/class-sensei-analysis.php:
```diff
@@ -320,190 +320,124 @@ public function analysis_page() {
-				<?php $sensei_analysis_lesson_users->display(); ?>
-			</div>
-			<div class="sensei-analysis-extra">
-				<?php do_action( 'sensei_analysis_extra' ); ?>
```
No usages found on GH.
Consider removing or deprecating?

m1r0 replied:
It's a really old hook (8+ years). I think it would be better to deprecate it first.

m1r0 commented:
Looks pretty good so far! One thing that I think could be improved a bit is the grading filters. Maybe remove the space between the inputs and have a filter button that is disabled by default and enabled once a lesson is selected?

merkushin committed:
Add method comments and fix other linter issues

merkushin committed:
Add param comments to fix linter issues

merkushin committed:
Fix grading filters styles

merkushin committed:
Escape link's URL

merkushin committed:
Fix default menu item

merkushin committed:
Move call of sensei_learners_extra action to the bottom extra tablenav

m1r0 reviewed (commented):
Hey Dmitry, thanks for improving this part! 🙏 

I have a few comments (mostly visual), so here it goes:

---

> I noticed that filters started adding unnecessary params like _wpnonce and _wp_http_referer to the query. These params we add to the form, the same on the Posts page. But when I submit filters for Posts it doesn't add mentioned params to the query. Can't find how it is filtered there. Any ideas?

WP core is [removing the parameters and making a redirect](https://github.com/WordPress/WordPress/blob/5dc1150ea2bbb3194805c14dcde055fc64be212b/wp-admin/edit.php#L231), but I'm not entirely sure why, to be honest. 🙂 Maybe it's a security issue? How hard would it be to do the same?

---

On mobile (chrome), the search input is missaligned.

![image](https://user-images.githubusercontent.com/1612178/178599295-b13360fc-56a5-4aa4-845c-338a06eda227.png)
![image](https://user-images.githubusercontent.com/1612178/178599328-6a7915de-b464-4bcd-badd-194b8f312d92.png)

---

The height of the select (select2) inputs is different than the button. I think it was like this before, but could it be easily fixed at least for the filters?

![image](https://user-images.githubusercontent.com/1612178/178599957-b0776e1c-2b41-476a-88be-40ee9f312882.png)

---

On mobile, the Grading filter inputs/button look small compared to the other inputs/buttons.  The "Reset filter" button also has too much white space at the end.

Grading filters:
![image](https://user-images.githubusercontent.com/1612178/178601388-8ea4f3d4-e08d-43ec-8219-565d12f47b1f.png)

Other filters:
![image](https://user-images.githubusercontent.com/1612178/178601700-2229680d-41d0-47a9-9c17-3f40869ddee9.png)

---

On mobile, the course filter on Reports -> Lessons is a bit off compared to the filter button.

![image](https://user-images.githubusercontent.com/1612178/178602624-4698ba82-94f8-42c9-aa0a-2dcd8ebfc8d8.png)

---

If you feel some of the mentions are not related to the current PR, I'm OK with opening separate issues for them. 👍

m1r0 commented on includes/admin/class-sensei-learners-admin-bulk-actions-view.php:
```diff
@@ -380,20 +408,18 @@ public function data_table_header() {
+								continue;
 							}
-							$this->courses_select( $courses, $selected_course, 'courses-select-filter', 'filter_by_course_id', __( 'Filter By Course', 'sensei-lms' ) );
-							?>
-							<button type="submit" id="filt" class="button action"><?php echo esc_html__( 'Filter', 'sensei-lms' ); ?></button>
-						</div>
-					</form>
+							echo '<input type="hidden" name="' . esc_attr( $name ) . '" value="' . esc_attr( $value ) . '">';
+						}
```
Hm, this causes input duplication now that we have one form.

![image](https://user-images.githubusercontent.com/1612178/178611474-37a4c6c8-045e-43ff-890d-056ff2bf903c.png)

I've also noticed the same issue on the other screens.

![image](https://user-images.githubusercontent.com/1612178/178611817-85c82a48-622a-4f83-84fe-2f3be83af10a.png)

I don't think it's causing issues, so I'm also ok with leaving it as-is for now.

merkushin committed:
Exclude hidden fields from filters that we displayed earlier

merkushin committed:
Exclude hidden fields from filters that we displayed earlier on Students page

merkushin committed:
Fix overlapping of filters in mobile view

merkushin committed:
Escape attribute value

merkushin committed:
Align select a course select with the submit button

merkushin committed:
Align select2 height with the button

merkushin commented:
@m1r0 

> WP core is [removing the parameters and making a redirect](https://github.com/WordPress/WordPress/blob/5dc1150ea2bbb3194805c14dcde055fc64be212b/wp-admin/edit.php#L231), but I'm not entirely sure why, to be honest. 🙂 Maybe it's a security issue? How hard would it be to do the same?

Haha, eventually, I know the answer. Yep, I think the main issue with these parameters is security. I suppose it is not safe to share your link with them, for example. I'm going to add the same redirect on our pages.

merkushin committed:
Merge branch 'trunk' into update/usage-of-list-table

# Conflicts:
#	includes/reports/overview/list-table/class-sensei-reports-overview-list-table-abstract.php

merkushin committed:
Don't output nonce and referer in sensei list tables

We can't redirect from these pages as output has already started. At the same time, we don't have POST forms and we can omit those fields.

merkushin committed:
Fix minor code style issues

merkushin committed:
Fix indentation

merkushin commented:
@m1r0 Could you take a look again?
I didn't fix styles for a filter from this [screenshot](https://user-images.githubusercontent.com/1612178/178601388-8ea4f3d4-e08d-43ec-8219-565d12f47b1f.png). It is completely custom, I think it would be better to work on it in a separate task.
I think I didn't remove all duplications of hidden fields, but, as you noticed, it doesn't break anything.
The problem with nonce and referrer fields I solved by [not displaying](https://github.com/Automattic/sensei/pull/5174/commits/2d0c8a794cf476497a8fccaadb1c5e366a8a3069) them :) We don't have POST forms on these pages, so I consider it safe. Unfortunately, we can't redirect on these pages, because output had already started by that time.

@Imran92 Feel free to join if you have some time :)

m1r0 commented:
The changes look good! Just one more thing I've noticed is the search on the reports working funky. Maybe it's related to the inputs?

https://user-images.githubusercontent.com/1612178/182258313-0feb41f6-d572-4207-aade-f90f0afc89c3.mp4

merkushin committed:
Don't duplicate hidden search field

merkushin commented:
Thanks @m1r0!
Fixed the issue here: https://github.com/Automattic/sensei/pull/5174/commits/65c173f9ad5f0ec16cea8118bcffdfe7ffd3f580

For some reason, I started "filtering" `a` field instead of `s` (for search). 
So it appeared with the previous value later on the page and caused that effect.

m1r0 reviewed (approved):
Looks good and works well. Thank you!

fjorgemota commented:
Just reporting out to keep note: I found that this PR caused the issue on #5504, which, curiously, only happens when there are no items on the list page (in this case, for the students page).

I'm still investigating the optimal way to fix this, but apparently, the root cause is that this PR opens a big `<form>` container (which has `method="GET"`) on the table, as per 1cf3d7bfdbc0d4b376a81a70b9101204c4050e65, and this causes an issue because `<form>` tags can't be nested, so technically each time we open a `<form>` tag inside that tag, the opening that is ignored, but the end tag isn't (well, kinda like that), so that causes some annoying HTML validation errors (according to Firefox source code view), which then causes the error...

Just sharing because it might be another thing to consider when doing those sorts of refactorings, especially when considering HTML forms. :)

**UPDATE:** I sent PR https://github.com/Automattic/sensei/pull/5583 to fix the new issue.
